File: android/src/main/java/com/margelo/nitro/rive/HybridRiveView.kt
package com.margelo.nitro.rive

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.core.Promise
import com.rive.BindData
import com.rive.RiveReactNativeView
import com.rive.ViewConfiguration
import app.rive.runtime.kotlin.core.Fit as RiveFit
import app.rive.runtime.kotlin.core.Alignment as RiveAlignment
import app.rive.runtime.kotlin.core.errors.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

fun Variant_HybridViewModelInstanceSpec_DataBindMode_DataBindByName?.toBindData(): BindData {
  if (this == null) return BindData.Auto

  return when (this) {
    is Variant_HybridViewModelInstanceSpec_DataBindMode_DataBindByName.First -> {
      val instance = (this.asFirstOrNull() as? HybridViewModelInstance)?.viewModelInstance
        ?: throw Error("Invalid ViewModelInstance")
      BindData.Instance(instance)
    }
    is Variant_HybridViewModelInstanceSpec_DataBindMode_DataBindByName.Second -> {
      when (this.asSecondOrNull()) {
        DataBindMode.AUTO -> BindData.Auto
        DataBindMode.NONE -> BindData.None
        else -> BindData.None
      }
    }
    is Variant_HybridViewModelInstanceSpec_DataBindMode_DataBindByName.Third -> {
      val name = this.asThirdOrNull()?.byName ?: throw Error("Missing byName value")
      BindData.ByName(name)
    }
  }
}

object DefaultConfiguration {
  const val AUTOPLAY = true
  val FIT = RiveFit.CONTAIN
  val ALIGNMENT = RiveAlignment.CENTER
  val LAYOUTSCALEFACTOR = null
}

@Keep
@DoNotStrip
class HybridRiveView(val context: ThemedReactContext) : HybridRiveViewSpec() {
  companion object {
    private const val TAG = "HybridRiveView"
  }

  //region State
  override val view: RiveReactNativeView = RiveReactNativeView(context)
  private var needsReload = false
  private var dataBindingChanged = false
  private var initialUpdate = true
  private var registeredFile: HybridRiveFile? = null
  //endregion

  //region View Props
  override var artboardName: String? = null
    set(value) {
      changed(field, value) { field = it }
    }
  override var stateMachineName: String? = null
    set(value) {
      changed(field, value) { field = it }
    }
  override var autoPlay: Boolean? = null
    set(value) {
      changed(field, value) { field = it }
    }
  override var file: HybridRiveFileSpec = HybridRiveFile()
    set(value) {
      if (field != value) {
        registeredFile?.unregisterView(this)
        registeredFile = null
      }
      changed(field, value) { field = it }
    }
  override var alignment: Alignment? = null
  override var fit: Fit? = null
  override var layoutScaleFactor: Double? = null
  override var dataBind: Variant_HybridViewModelInstanceSpec_DataBindMode_DataBindByName? = null
    set(value) {
      if (field != value) {
        field = value
        dataBindingChanged = true
      }
    }
  override var onError: (error: RiveError) -> Unit = {}
  //endregion

  //region View Methods
  override fun awaitViewReady(): Promise<Boolean> {
    return Promise.async {
      withContext(Dispatchers.Main) {
        view.awaitViewReady()
      }
    }
  }

  override fun bindViewModelInstance(viewModelInstance: HybridViewModelInstanceSpec) =
    executeOnUiThread {
      val hybridVmi = viewModelInstance as? HybridViewModelInstance ?: return@executeOnUiThread;
      view.bindViewModelInstance(hybridVmi.viewModelInstance)
    }

  override fun getViewModelInstance(): HybridViewModelInstanceSpec? {
    val viewModelInstance = view.getViewModelInstance() ?: return null
    return HybridViewModelInstance(viewModelInstance)
  }

  override fun play() = executeOnUiThread { view.play() }

  override fun pause() = executeOnUiThread { view.pause() }

  override fun onEventListener(onEvent: (event: UnifiedRiveEvent) -> Unit) =
    executeOnUiThread { view.addEventListener(onEvent) }

  override fun removeEventListeners() = executeOnUiThread { view.removeEventListeners() }

  override fun setNumberInputValue(name: String, value: Double, path: String?) =
    view.setNumberInputValue(name, value, path)

  override fun getNumberInputValue(name: String, path: String?): Double =
    view.getNumberInputValue(name, path)

  override fun setBooleanInputValue(name: String, value: Boolean, path: String?) =
    view.setBooleanInputValue(name, value, path)

  override fun getBooleanInputValue(name: String, path: String?): Boolean =
    view.getBooleanInputValue(name, path)

  override fun triggerInput(name: String, path: String?) = view.triggerInput(name, path)

  override fun setTextRunValue(name: String, value: String, path: String?) =
    view.setTextRunValue(name, value, path)

  override fun getTextRunValue(name: String, path: String?): String =
    view.getTextRunValue(name, path)
  //endregion

  //region Update
  fun refreshAfterAssetChange() {
    afterUpdate()
  }


  override fun afterUpdate() {
    logged(TAG, "afterUpdate") {
      val hybridFile = file as? HybridRiveFile
      val riveFile = hybridFile?.riveFile ?: return@logged

      val config = ViewConfiguration(
        artboardName = artboardName,
        stateMachineName = stateMachineName,
        autoPlay = autoPlay ?: DefaultConfiguration.AUTOPLAY,
        riveFile = riveFile,
        alignment = convertAlignment(alignment) ?: DefaultConfiguration.ALIGNMENT,
        fit = convertFit(fit) ?: DefaultConfiguration.FIT,
        layoutScaleFactor = layoutScaleFactor?.toFloat() ?: DefaultConfiguration.LAYOUTSCALEFACTOR,
        bindData = dataBind.toBindData()
      )
      view.configure(config, dataBindingChanged=dataBindingChanged, needsReload, initialUpdate= initialUpdate)

      if (needsReload && hybridFile != null) {
        hybridFile.registerView(this)
        registeredFile = hybridFile
      }

      needsReload = false
      dataBindingChanged = false
      initialUpdate = false
      super.afterUpdate()
    }
  }
  //endregion

  //region Helpers
  private fun <T> changed(current: T, new: T, setter: (T) -> Unit) {
    if (current != new) {
      setter(new)
      needsReload = true
    }
  }

  private fun executeOnUiThread(action: () -> Unit) {
    context.currentActivity?.runOnUiThread() {
      try {
        action()
      } catch (e: Exception) {
        throw Error(e.message) // TODO: Correctly handling errors (https://nitro.margelo.com/docs/errors)
      } catch (e: Error) {
        throw Error(e.message)
      }
    }
  }

  private fun convertAlignment(alignment: Alignment?): RiveAlignment? {
    if (alignment == null) return null

    return when (alignment) {
      Alignment.TOPLEFT -> RiveAlignment.TOP_LEFT
      Alignment.TOPCENTER -> RiveAlignment.TOP_CENTER
      Alignment.TOPRIGHT -> RiveAlignment.TOP_RIGHT
      Alignment.CENTERLEFT -> RiveAlignment.CENTER_LEFT
      Alignment.CENTER -> RiveAlignment.CENTER
      Alignment.CENTERRIGHT -> RiveAlignment.CENTER_RIGHT
      Alignment.BOTTOMLEFT -> RiveAlignment.BOTTOM_LEFT
      Alignment.BOTTOMCENTER -> RiveAlignment.BOTTOM_CENTER
      Alignment.BOTTOMRIGHT -> RiveAlignment.BOTTOM_RIGHT
    }
  }

  private fun convertFit(fit: Fit?): RiveFit? {
    if (fit == null) return null

    return when (fit) {
      Fit.FILL -> RiveFit.FILL
      Fit.CONTAIN -> RiveFit.CONTAIN
      Fit.COVER -> RiveFit.COVER
      Fit.FITWIDTH -> RiveFit.FIT_WIDTH
      Fit.FITHEIGHT -> RiveFit.FIT_HEIGHT
      Fit.NONE -> RiveFit.NONE
      Fit.SCALEDOWN -> RiveFit.SCALE_DOWN
      Fit.LAYOUT -> RiveFit.LAYOUT
    }
  }

  private fun detectErrorType(exception: Exception): Pair<RiveErrorType, String> {
    val message = exception.message ?: exception.toString()
    val type = when (exception) {
      is ArtboardException -> RiveErrorType.INCORRECTARTBOARDNAME
      is StateMachineException -> RiveErrorType.INCORRECTSTATEMACHINENAME
      is AnimationException -> RiveErrorType.UNKNOWN
      is MalformedFileException -> RiveErrorType.MALFORMEDFILE
      is StateMachineInputException -> RiveErrorType.INCORRECTSTATEMACHINEINPUTNAME
      is TextValueRunException -> RiveErrorType.UNKNOWN
      is ViewModelException -> RiveErrorType.VIEWMODELINSTANCENOTFOUND
      else -> RiveErrorType.UNKNOWN
    }
    return Pair(type, message)
  }

  fun logged(tag: String, note: String? = null, fn: () -> Unit) {
    try {
      fn()
    } catch (e: Exception) {
      val (errorType, errorDescription) = detectErrorType(e)
      val noteString = note?.let { " $it" } ?: ""
      val errorMessage = "[RIVE] $tag$noteString $errorDescription"
      val riveError = RiveError(
        type = errorType,
        message = errorMessage
      )
      onError(riveError)
    }
  }
  //endregion
}

File: ios/HybridRiveView.swift
import Foundation
import NitroModules
import RiveRuntime
import UIKit

private struct DefaultConfiguration {
  static let autoPlay = true
  static let alignment = RiveAlignment.center
  static let fit = RiveFit.contain
  static let layoutScaleFactor = RiveRuntime.RiveViewModel.layoutScaleFactorAutomatic
}

typealias HybridDataBindMode = Variant__any_HybridViewModelInstanceSpec__DataBindMode_DataBindByName

extension Optional
where Wrapped == HybridDataBindMode {
  func toDataBingMode() throws -> BindData {
    guard let value = self else {
      return .auto
    }

    switch value {
    case .first(let viewModelInstance):
      if let instance = (viewModelInstance as? HybridViewModelInstance)?.viewModelInstance {
        return .instance(instance)
      } else {
        throw RuntimeError.error(withMessage: "Invalid ViewModelInstance")
      }
    case .second(let mode):
      switch mode {
      case .auto:
        return .auto
      case .none:
        return .none
      }
    case .third(let dataBindByName):
      return .byName(dataBindByName.byName)
    }
  }
}

class HybridRiveView: HybridRiveViewSpec {
  // MARK: View Props
  var dataBind: HybridDataBindMode? = nil {
    didSet {
      dataBindingChanged = true
    }
  }

  var artboardName: String? { didSet { needsReload = true } }
  var stateMachineName: String? { didSet { needsReload = true } }
  var autoPlay: Bool? { didSet { needsReload = true } }
  var file: (any HybridRiveFileSpec) = HybridRiveFile() {
    didSet { needsReload = true }
  }
  var alignment: Alignment?
  var fit: Fit?
  var layoutScaleFactor: Double?
  var onError: (RiveError) -> Void = { _ in }

  func awaitViewReady() throws -> Promise<Bool> {
    return Promise.async { [self] in
      return try await getRiveView().awaitViewReady()
    }
  }

  func bindViewModelInstance(viewModelInstance: (any HybridViewModelInstanceSpec)) throws {
    guard let viewModelInstance = (viewModelInstance as? HybridViewModelInstance)?.viewModelInstance
    else { return }
    try getRiveView().bindViewModelInstance(viewModelInstance: viewModelInstance)
  }

  func getViewModelInstance() throws -> (any HybridViewModelInstanceSpec)? {
    guard let viewModelInstance = try getRiveView().getViewModelInstance() else { return nil }
    return HybridViewModelInstance(viewModelInstance: viewModelInstance)
  }

  func play() throws { try getRiveView().play() }

  func pause() throws { try getRiveView().pause() }

  func onEventListener(onEvent: @escaping (UnifiedRiveEvent) -> Void) throws {
    try getRiveView().addEventListener(onEvent)
  }

  func removeEventListeners() throws { try getRiveView().removeEventListeners() }

  func setNumberInputValue(name: String, value: Double, path: String?) throws {
    try getRiveView().setNumberInputValue(name: name, value: Float(value), path: path)
  }

  func getNumberInputValue(name: String, path: String?) throws -> Double {
    return try Double(getRiveView().getNumberInputValue(name: name, path: path))
  }

  func setBooleanInputValue(name: String, value: Bool, path: String?) throws {
    try getRiveView().setBooleanInputValue(name: name, value: value, path: path)
  }

  func getBooleanInputValue(name: String, path: String?) throws -> Bool {
    return try getRiveView().getBooleanInputValue(name: name, path: path)
  }

  func triggerInput(name: String, path: String?) throws {
    try getRiveView().triggerInput(name: name, path: path)
  }

  func setTextRunValue(name: String, value: String, path: String?) throws {
    try getRiveView().setTextRunValue(name: name, value: value, path: path)
  }

  func getTextRunValue(name: String, path: String?) throws -> String {
    return try getRiveView().getTextRunValue(name: name, path: path)
  }

  // MARK: Views
  var view: UIView = RiveReactNativeView()
  func getRiveView() throws -> RiveReactNativeView {
    guard let riveView = view as? RiveReactNativeView else {
      throw RuntimeError.error(withMessage: "RiveReactNativeView is null or not configured")
    }
    return riveView
  }

  // MARK: Update
  func afterUpdate() {
    logged(tag: "HybridRiveView", note: "afterUpdate") {
      guard let hybridFile = file as? HybridRiveFile,
        let file = hybridFile.riveFile
      else { return }

      let config = ViewConfiguration(
        artboardName: artboardName,
        stateMachineName: stateMachineName,
        autoPlay: autoPlay ?? DefaultConfiguration.autoPlay,
        riveFile: file,
        viewSource: hybridFile,
        alignment: convertAlignment(alignment) ?? DefaultConfiguration.alignment,
        fit: convertFit(fit) ?? DefaultConfiguration.fit,
        layoutScaleFactor: layoutScaleFactor ?? DefaultConfiguration.layoutScaleFactor,
        bindData: try dataBind.toDataBingMode()
      )

      let riveView = try getRiveView()
      try riveView.configure(
        config, dataBindingChanged: dataBindingChanged, reload: needsReload,
        initialUpdate: initialUpdate)
      needsReload = false
      dataBindingChanged = false
      initialUpdate = false
    }
  }

  // MARK: Internal State
  private var needsReload = false
  private var dataBindingChanged = false
  private var initialUpdate = true

  // MARK: Helpers
  private func convertAlignment(_ alignment: Alignment?) -> RiveAlignment? {
    guard let alignment = alignment else { return nil }

    switch alignment {
    case .topleft: return .topLeft
    case .topcenter: return .topCenter
    case .topright: return .topRight
    case .centerleft: return .centerLeft
    case .center: return .center
    case .centerright: return .centerRight
    case .bottomleft: return .bottomLeft
    case .bottomcenter: return .bottomCenter
    case .bottomright: return .bottomRight
    }
  }

  private func convertFit(_ fit: Fit?) -> RiveFit? {
    guard let fit = fit else { return nil }

    switch fit {
    case .fill: return .fill
    case .contain: return .contain
    case .cover: return .cover
    case .fitwidth: return .fitWidth
    case .fitheight: return .fitHeight
    case .none: return .noFit
    case .scaledown: return .scaleDown
    case .layout: return .layout
    }
  }
}

extension HybridRiveView {
  func logged(tag: String, note: String? = nil, _ fn: () throws -> Void) {
    do {
      return try fn()
    } catch (let e) {
      let errorType = detectErrorType(e)
      let errorMessage = "[RIVE] \(tag) \(note ?? "") \(e)"

      let riveError = RiveError(
        message: errorMessage,
        type: errorType
      )
      onError(riveError)
    }
  }

  private func detectErrorType(_ error: Error) -> RiveErrorType {
    if case NativeRiveError.instanceNotFound = error {
      return .databindingerror
    }

    let nsError = error as NSError
    guard let errorName = nsError.userInfo["name"] as? String else {
      return .unknown
    }

    switch errorName {
    case "NoArtboardFound":
      return .incorrectartboardname
    case "NoStateMachineFound":
      return .incorrectstatemachinename
    case "NoAnimationFound":
      return .incorrectanimationname
    case "Malformed":
      return .malformedfile
    case "FileNotFound":
      return .filenotfound
    case "NoStateMachineInputFound":
      return .incorrectanimationname
    case "TextRunNotFoundError":
      return .textrunnotfounderror
    case "DataBindingError":
      return .databindingerror
    default:
      return .unknown
    }
  }
}
